hack/generator/pkg/conversions/bag_item_type.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package conversions
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/dave/dst"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+)
+
+// BagItemType is a wrapper type for properties that are stored-in/recalled-from a PropertyBag
+type BagItemType struct {
```

Comment:
I did initially spike using flags or some other special context to carry this information through, and (essentially) the whole thing disolved into a mass of special and edge cases. Using a custom type made things fall out with a whole lot less code and made it much simpler.

Reply:
I can sorta squint and see the type-ness of "a type which has been serialized to JSON or YAML and stored in a string" as distinct from the type that was serialized. When looking at it that way, I can see how it fits into the model you have here where source and destination `astmodel.Type` are the primary keys for picking what conversion is run.

So `Serialize(ObjectType) -> BagItemType<ObjectType>` and `Deserialize(BagItemType<ObjectType>) -> ObjectType`. It might have helped me if in the description or comment there was a bit more explaination about what this is from a `Type` perspective - it is a flag or wrapper but that flag/wrapper does kinda have a "logical meaning" too, if that makes sense?

In that vein, would another name for this be something like `SerializedType`? (not necessarily suggesting you change, just making sure I understand how you're seeing it/thinking about it)

Also, given that the intent here is to just use this for to/from the `PropertyBag`, I think it makes sense to `panic` in `AsType`, similar to how we do in `AllOfType`? Possibly in all of the following methods actually?
1. AsType
2. AsDeclarations (you already panic here)
3. AsZero
4. RequiredPackageReferences
5. References (AllOfType doesn't panic here, so there may be a good reason for you to avoid it too?)

This is an "ephemeral" type that is used for internal conversions only and shouldn't ever be rendered (or attempted to render) to a file.

Reply:
I think you're seeing differences that I'm not. 

An endpoint with `Optional<string>` is saying "this string might be missing, the generated code needs to handle that", and is picked up by the appropriate handler, with different cases for read vs write.

Similarly, an endpoint with `BagItem<string>` is saying "this string doesn't get stored in a property" and it picked up by the appropriate handler, with different cases for read vs write.

Reply:
Also, I implemented the methods that were necessary for the code generation to work - reducing the `BagItem` by forcing all those methods to panic would make the rest of the code more complicated, for (IMHO) no net gain.

Reply:
I missed their usage in `pullFromBagItem` in `property_conversions.go`.

The thing that I struggled with w/ the `​BagItemType`​ was trying to understand what construct it represented in go code (or in the "logical type structure" we've built to represent go code and other related constructs). I understand what a ​`PrimitiveType​​` or an `ArrayType` represent, how they compose, and what operations can take me from one to another. I was thinking that `BagItemType` would fit into that same framework and so was trying to understand what it renders into in code, or what it represents and was having a hard time doing that.

Admittedly, we have a few other `Type`'s which don't follow those rules (`FlaggedType` mostly, as I can't really tell you what that represents either). Speaking of `FlaggedType`, if all `BagItemType` is for is smuggling some metadata around, and it really just defers to its `Element` for everything, why not just use `FlaggedType` which also does that?  You could just define a custom `TypeFlag` in this package and use it on `FlaggedType` and probably save yourself this whole type?

Reply:
_Maaayyyybeeee ..._

Would you mind if I deferred investigating that to a later branch/PR?